### PR TITLE
tests: IATR E2E tests

### LINKED ISSUE #25351: [IATR] E2E Tests for IATR
When working through the IATR M0 and M1 stories, the focus was on writing component tests for the new front end features and updating existing end-to-end tests.  To complete the milestone, the CT tests need to be supported with E2E tests focused on the new functionality.  

These tests should be branched off the `stokes/IATR-M0-cloud-run-data` branch for the https://github.com/cypress-io/cypress/pull/25316 PR unless it is been merged to the feature branch `feature/IATR-M0`

The following areas need to be covered:

* Sidebar
  * See https://github.com/cypress-io/cypress/pull/25152
* Debug page
  * Empty states
    * These are similar if not the same as the Runs page.  So scenarios can be copied from there
    * See https://github.com/cypress-io/cypress/pull/25219
  * Non failure run statuses
  	* See https://github.com/cypress-io/cypress/pull/25286
  	* See https://github.com/cypress-io/cypress/pull/25361
  * Failed build showing failed specs
	* Failed build with only one group
  	* Failed build with multiple groups
  	* See https://github.com/cypress-io/cypress/pull/25263
  * Banners shown for "next" run
    * Completed state with running next state
    * Completed state with running next state that completes while on the page
    * See https://github.com/cypress-io/cypress/pull/25282
  * Navigating to external link for Cypress Cloud
  * Using the "Run failures button"
    * Opening test runner to run failed tests
    * Opening tooltip to change test type
    * See https://github.com/cypress-io/cypress/pull/25234

### PR DESCRIPTION
**Pull request by lmiller1990**
- Closes https://github.com/cypress-io/cypress/issues/25351

### Additional details

Adds End to End tests for the Debug page with some useful fixtures. I am not sure how maintainable these fixtures are, but this is the only way I could reliably get the app into the correct state for testing. I generated these from a real project, and copy and pasted them in.

I can make the fixtures more minimal if we think that'd be useful; my main goal here was actually to capture things in a real world fashion... the existing fixtures incredibly difficult to understand and debug, since they are all generated programatically. A lot of time was wasted trying to reason about them.

I'm going to make some suggestions and recommendations moving forward; we need more real world, production like data sets to write End to End tests against. The style I'm using here is verbose, but it is SIMPLE - just look at the test, all the info you need is right there - just look at the fixtures, which are static JSON - vastly easier to debug.

### Steps to test

Run the End to End tests, if you want!

### How has the user experience changed?

Tests

### PR Tasks

- [x] Have tests been added/updated?
- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/app/cypress/fixtures/gql-Debug.json [hunk 1]
```diff
@@ -0,0 +1,180 @@
+{
```
Comment (lmiller1990) [inline packages/app/cypress/fixtures/gql-Debug.json]: Concerned about the maintainability of these fixtures going forward.

File: packages/app/src/store/run-all-specs-store.ts [hunk 2]
```diff
@@ -77,6 +76,8 @@ export const useRunAllSpecsStore = defineStore('runAllSpecs', () => {
     directoryChildrenRef.value = directoryChildren
   }
 
+  const query = useQuery({ query: RunAllSpecsDataDocument, pause: isRunMode || window.__CYPRESS_TESTING_TYPE__ === 'component' })
```
Comment (lmiller1990) [inline packages/app/src/store/run-all-specs-store.ts:79]: Run All Specs is E2E only, drive-by optimization - we don't need to query if we are in CT mode.

File: packages/frontend-shared/cypress/e2e/e2ePluginSetup.ts [hunk 3]
```diff
@@ -349,8 +351,12 @@ async function makeE2ETasks () {
       return null
     },
 
-    __internal_remoteGraphQLIntercept (fn: string) {
-      remoteGraphQLIntercept = new Function('console', 'obj', 'testState', `return (${fn})(obj, testState)`).bind(null, console) as RemoteGraphQLInterceptor
+    __internal_remoteGraphQLIntercept (args: {
```
Comment (lmiller1990) [inline packages/frontend-shared/cypress/e2e/e2ePluginSetup.ts:354]: This is so we can pass data from the frontend to `remoteGraphQLIntercept`

That's what I do here https://github.com/cypress-io/cypress/pull/25601/files#diff-b0772ecf528a07f437bcf6dc261b015b048d3403c7cecc62a3b18823cc891a37R28

File: packages/app/cypress/e2e/debug.cy.ts [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+      cy.contains('src/components/InfoPanel/InfoPanel.cy.ts')
+      cy.findByTestId('metaData-Results-spec-duration').contains('00:04')
+      cy.findByTestId('metaData-Results-operating-system').contains('Linux Ubuntu')
+      cy.findByTestId('metaData-Results-browser').contains('Electron 106')
+      cy.findByTestId('metaData-Results-testing-type').contains('Component')
+    })
+
+    cy.findByTestId('test-row').contains('InfoPanel')
+    cy.findByTestId('test-row').contains('renders')
+    cy.findByTestId('run-failures').should('exist').should('have.attr', 'href', '#/specs/runner?file=src/components/InfoPanel/InfoPanel.cy.ts&mode=debug')
```
Comment (astone123) [inline packages/app/cypress/e2e/debug.cy.ts:149]: Should we add assertions here for the hover actions in the test row?

File: packages/app/cypress/e2e/support/e2eSupport.ts [hunk 5]
```diff
@@ -2,6 +2,14 @@ import '@packages/frontend-shared/cypress/support/e2e'
 import 'cypress-real-events/support'
 import './execute-spec'
 
+Cypress.on('window:before:load', (win) => {
+  // Can set this in a spec-by-spec basis to 'true' to use
```
Comment (astone123) [inline packages/app/cypress/e2e/support/e2eSupport.ts:6]: Nice comment here 👍🏻

File: packages/app/cypress/e2e/debug.cy.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    cy.openProject('cypress-in-cypress')
+    cy.startAppServer('component')
+
+    cy.loginUser()
+    cy.withCtx((ctx) => {
+      ctx.git?.__setGitHashesForTesting(['commit1', 'commit2'])
+    })
+
+    cy.remoteGraphQLIntercept((obj, _testState, options) => {
+      if (obj.operationName === 'RelevantRunsDataSource_RunsByCommitShas') {
```
Comment (astone123) [inline packages/app/cypress/e2e/debug.cy.ts:23]: Why do we need to check for this query specifically when it looks like the other ones can be intercepted with `cy.intercept` since we're not using websockets during the test?

**Review by astone123 — APPROVED** [review]
Wow, lots of setup to mock these responses. Nice job getting that working. 

I wonder if we should add a test case for each of the [empty states](https://www.figma.com/file/PAOozuCxtfetqP8fUVvg0y/branch/QkCp2F69X0qyLo7qnYQ67e/Cypress-App?t=v2XdR7vR6PGIfHk4-0) (not logged in, project not connected, no runs, etc) or if that's overkill for this PR.

I left a few comments, mostly questions. Either way, this looks like a good starting point for writing e2e tests for the later iterations of the debug feature

File: packages/app/cypress/e2e/debug.cy.ts (hunk 6 above)
Comment (warrensplayer) [reply packages/app/cypress/e2e/debug.cy.ts:23]: I have the same question as Adam. Is this needed because it fires before the setting of the `__CYPRESS_GQL_NO_SOCKET__` ?  

I do like the ability to pass the fixtures through to the `removeGraphQLIntercept`

**Review by warrensplayer — APPROVED** [review]

File: packages/app/cypress/e2e/debug.cy.ts (hunk 6 above)
Comment (lmiller1990) [reply packages/app/cypress/e2e/debug.cy.ts:23]: I will bring this up at the "graphql e2e paint points" meeting, but basically `cy.intercept` is only for requests from the **front end** via `fetch`.

This means `cy.intercept` cannot work with subscriptions, those are always over websockets. What happens here is

1. subscription (via websocket)
2. this triggers a remote graphql request on the server

The only way to control the response for that is to use `remoteGraphQLIntercept` (this function is for intercepting request on the **server**). Despite the name, this is something we implemented specifically for GraphQL and it does **not** match the same signature as `cy.intercept` (but it could, and probably should).

**Comment by lmiller1990** [comment]
> Wow, lots of setup to mock these responses. Nice job getting that working.
> 
> I wonder if we should add a test case for each of the [empty states](https://www.figma.com/file/PAOozuCxtfetqP8fUVvg0y/branch/QkCp2F69X0qyLo7qnYQ67e/Cypress-App?t=v2XdR7vR6PGIfHk4-0) (not logged in, project not connected, no runs, etc) or if that's overkill for this PR.
> 
> I left a few comments, mostly questions. Either way, this looks like a good starting point for writing e2e tests for the later iterations of the debug feature

Yep we should, now that this is set up it should be pretty quick to add those - although I'm less concerned about those e2e tests, the ones here give us about 90% of the coverage (but we should loop back for the others, too).
